elastic-apm-agent/src/main/java/co/elastic/apm/agent/premain/ShadedClassLoader.java

```diff
@@ -0,0 +1,139 @@
+
+public class ShadedClassLoader extends URLClassLoader {
+
+    private static final String CLASS_EXTENSION = ".class";
+    private final String prefix;
+    private final String classFileExtension;
+    private final Manifest manifest;
+    private final URL agentJarUrl;
+
+    static {
+        ClassLoader.registerAsParallelCapable();
+    }
+
+    public ShadedClassLoader(File agentJar, ClassLoader parent, String prefix, String classNameExtension) throws IOException {
+        super(new URL[]{agentJar.toURI().toURL()}, parent);
+        this.prefix = prefix;
+        this.classFileExtension = classNameExtension;
+        this.agentJarUrl = agentJar.toURI().toURL();
+        try (JarFile jarFile = new JarFile(agentJar)) {
+            this.manifest = jarFile.getManifest();
+        }
+    }
+
+    @Override
+    protected Class<?> findClass(String name) throws ClassNotFoundException {
+        try {
+            return super.findClass(name);
+        } catch (ClassNotFoundException e) {
+            byte[] classBytes = getClassBytes(name);
+            if (classBytes != null) {
+                return defineClass(name, classBytes);
+            } else {
+                throw new ClassNotFoundException(name);
+            }
+        }
+    }
+
+    private Class<?> defineClass(String name, byte[] classBytes) {
+        String packageName = getPackageName(name);
+        if (packageName != null && getPackage(packageName) == null) {
+            synchronized (this) {
+                if (getPackage(packageName) == null) {
+                    if (manifest != null) {
+                        definePackage(name, manifest, agentJarUrl);
+                    } else {
+                        definePackage(packageName, null, null, null, null, null, null, null);
+                    }
+                }
+            }
+        }
+        return defineClass(name, classBytes, 0, classBytes.length, ShadedClassLoader.class.getProtectionDomain());
+    }
+
+    @Nullable
+    public String getPackageName(String className) {
+        int i = className.lastIndexOf('.');
+        if (i != -1) {
+            return className.substring(0, i);
+        }
+        return null;
+    }
+
+    private byte[] getClassBytes(String name) throws ClassNotFoundException {
+        InputStream classStream = getResourceAsStream(prefix + name.replace('.', '/') + classFileExtension);
+        if (classStream != null) {
+            try (BufferedInputStream is = new BufferedInputStream(classStream)) {
+                ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+
+                int n;
+                byte[] data = new byte[4 * 1024];
+                while ((n = is.read(data, 0, data.length)) != -1) {
+                    buffer.write(data, 0, n);
+                }
+                return buffer.toByteArray();
+            } catch (IOException e) {
+                throw new ClassNotFoundException(name, e);
+            }
+        }
```

Comment:
Maybe we can reuse `java.security.SecureClassLoader#defineClass(java.lang.String, java.nio.ByteBuffer, java.security.CodeSource)` instead with a `ByteBuffer` and a `CodeSource` constructed with the class file URL?

Reply:
What would be the benefit?

Reply:
Ah, I think I see what you mean. Even with that overload of the method, you still need to load the class file bytes into the ByteBuffer. The CodeSource would contain the URL of the jar, not the class file.

Reply:
Sure, there is no way around reading those bytes, I thought about possible optimizations done natively and the ability to use non-heap memory for that. However, the `CodeSource` is not used for reading the bytes, only for the creation of the `ProtectionDomain`. 
I was only worried about possible additional load on the heap during class loading when seeing this 4K byte arrays. The OpenJDK implementation (`jdk.internal.loader.Resource#getBytes`) seems to prefer GC over usage of non-minimal byte arrays. So, without any evidence to real issue and only up to you, I propose to follow the same concept by reducing the byte array buffer size.

Reply:
I have reduced the buffer size to 1k and removed the BufferedInputStream